application/src/main/java/org/thingsboard/server/actors/device/DeviceActorMessageProcessor.java

```diff
@@ -705,10 +706,19 @@ private void processRpcResponseStatus(SessionInfoProto sessionInfo, ToDeviceRpcR
                 maxRpcRetries = maxRpcRetries == null ?
                         systemContext.getMaxRpcRetries() : Math.min(maxRpcRetries, systemContext.getMaxRpcRetries());
                 if (maxRpcRetries <= md.getRetries()) {
-                    toDeviceRpcPendingMap.remove(requestId);
-                    status = RpcStatus.FAILED;
-                    response = JacksonUtil.newObjectNode().put("error", "There was a Timeout and all retry " +
-                            "attempts have been exhausted. Retry attempts set: " + maxRpcRetries);
+                    if (closeTransportSessionOnRpcDeliveryTimeout) {
+                        md.setRetries(0);
+                        status = RpcStatus.QUEUED;
+                        sessions.forEach(this::notifyTransportAboutClosedSessionRpcDeliveryTimeout);
+                        attributeSubscriptions.clear();
+                        rpcSubscriptions.clear();
+                        dumpSessions();
```

Comment:
I see that this is duplicated, consider extracting to a method and refactor other usages